### PR TITLE
Remove +1 from newline detection

### PR DESCRIPTION
**Pull request by vjeux**
All the changes are related to spurious `;`. Sometimes the logic is more correct, sometimes less. Since `;` are going to be removed after the first save, I don't think it matters that much.

### CODE REVIEW

File: tests/async/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -133,7 +133,6 @@ async function foo() {
 // am not clear on whether it should. In any case it\'s a strange corner case
 // that is probably not important to support.
 class C {}
-
```
Comment (vjeux) [inline tests/async/__snapshots__/jsfmt.spec.js.snap]: This line was correct

```js
class C {};

var P: Promise<Class<C>> = new Promise(function (resolve, reject) {
```

File: tests/binding/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -412,7 +412,6 @@ function var_var() {
 function function_toplevel() {
   function a() {
   }
-
```
Comment (vjeux) [inline tests/binding/__snapshots__/jsfmt.spec.js.snap]: This line was not correct

```js
function function_toplevel() {
  function a() {};
  function a() {}; // OK
}
```

File: tests/declare_export/__snapshots__/jsfmt.spec.js.snap [hunk 3]
```diff
@@ -68,7 +68,6 @@ class Test extends Base {
     return 2;
   }
 }
-
```
Comment (vjeux) [inline tests/declare_export/__snapshots__/jsfmt.spec.js.snap]: This line was correct

```js
};

module.exports = Test;
```

File: tests/declare_export/__snapshots__/jsfmt.spec.js.snap [hunk 4]
```diff
@@ -181,7 +180,6 @@ declare export function getAFoo(): FooImpl;
  */
 
 declare export default class FooImpl { givesANum(): number }
-
```
Comment (vjeux) [inline tests/declare_export/__snapshots__/jsfmt.spec.js.snap]: This line was correct

```js
declare export default class FooImpl { givesANum(): number; };

// Regression test for https://github.com/facebook/flow/issues/511
```

File: tests/declare_export/__snapshots__/jsfmt.spec.js.snap [hunk 5]
```diff
@@ -204,7 +202,6 @@ declare export default class Foo { givesANum(): number; };
  */
 
 declare export default class Foo { givesANum(): number }
-
```
Comment (vjeux) [inline tests/declare_export/__snapshots__/jsfmt.spec.js.snap]: This line was not (this is the end of the file, there should be no spurious empty lines)

File: tests/declare_export/__snapshots__/jsfmt.spec.js.snap [hunk 6]
```diff
@@ -461,7 +458,6 @@ declare export { groupedSpecifierNumber1, groupedSpecifierNumber2 };
 
 declare export function givesANumber(): number;
 declare export class NumberGenerator { givesANumber(): number }
-
```
Comment (vjeux) [inline tests/declare_export/__snapshots__/jsfmt.spec.js.snap]: This one was correct

```js
declare export class NumberGenerator { givesANumber(): number; };

declare export var varDeclNumber1: number;
```

File: tests/enumerror/__snapshots__/jsfmt.spec.js.snap [hunk 7]
```diff
@@ -33,14 +33,12 @@ function isActive(
 ): boolean {
   return ad.state === \"ACTIVE\";
 }
-
```
Comment (vjeux) [inline tests/enumerror/__snapshots__/jsfmt.spec.js.snap]: This line was not correct

```js
    return ad.state === \'ACTIVE\';
};
isActive({state: \'PAUSE\'});
```

File: tests/enumerror/__snapshots__/jsfmt.spec.js.snap [hunk 8]
```diff
@@ -33,14 +33,12 @@ function isActive(
...
 isActive({ state: \"PAUSE\" });
 
 var MyStates = { PAUSED: \"PAUSED\", ACTIVE: \"ACTIVE\", DELETED: \"DELETED\" };
 function isActive2(ad: { state: $Keys<typeof MyStates> }): boolean {
   return ad.state === MyStates.ACTIVE;
 }
-
```
Comment (vjeux) [inline tests/enumerror/__snapshots__/jsfmt.spec.js.snap]: I can go on for the entire file but it's basically 50/50 correct and incorrect. I think we should not have the `+ 1` that doesn't make any sense and try to figure out why the newly incorrect cases here happen and fix them for real.

**Comment by jlongster** [comment]
I agree that we should know why something works. I dislike software where things happen to align and work the way you'd expect; if the results are unexpected let's figure it out. I'd really like to open a PR for binary expressions tonight and I'd like to take one last glance at the indexing logic to see if I can think of anything before merging this, so I'll wait until tomorrow morning to merge.

**Comment by vjeux** [comment]
Yeah, I also want to spend more time trying to figure out why the cases that don't work don't. But I figured I would send this PR to start the discussion :)

**Comment by jlongster** [comment]
I found the error and updated the code. I printed out the character that the `end` location points to and unfortunately, at least for babylon, it is inconsistent (I guess for Flow too since the tests changed).

If you had an expression `expr;\n`, sometimes `node.end` points to the `;` character and sometimes it points to `\n`. It depends on the kind of statement/expressions. It's not a difficult fix, but is unfortunate.

Do these tests updates look sane @vjeux ?

**Comment by vjeux** [comment]
superlikestamp

I went through 12 files and all the changes are correct! I think you fixed it for realz! Thanks :)
